### PR TITLE
Lecture 28: A Civil Engineering Perspective on AI

### PR DESCRIPTION
**Pull request by jmcohen**
https://jmcohen.github.io/pgm-spring-2019/notes/lecture-28/

### CODE REVIEW

File: _posts/2019-04-29-lecture-28.md [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+authors:
+  - name: Jiaxu Zou
+    url: "#"
+  - name: Amanda Coston
+    url: "#"
+  - name: Jeremy Cohen
+    url: "#"
+
+editors:
+  - name: TODO
```
Comment (alshedivat) [inline _posts/2019-04-29-lecture-28.md]:
```suggestion
  - name: Maruan Al-Shedivat
```

File: _posts/2019-04-29-lecture-28.md [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+    url: "#"  # optional URL to the editor's homepage
```
Comment (alshedivat) [inline _posts/2019-04-29-lecture-28.md]:
```suggestion
    url: "https://www.cs.cmu.edu/~mshediva/"
```

File: _posts/2019-04-29-lecture-28.md [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+---
+
+## Industrialization of AI
+### Barriers to AI Solutions Today
+According to a McKinsey CEO survey in 2018, only 39% of the companies made significant progress in implementing AI applications. Up to 7$ of the companies made no progress.
```
Comment (alshedivat) [inline _posts/2019-04-29-lecture-28.md]:
```suggestion
According to a McKinsey CEO survey in 2018, only 39% of the companies made significant progress in implementing AI applications. Up to 7% of the companies made no progress.
```

File: _posts/2019-04-29-lecture-28.md [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+then computes model updates using its own subset of the dataset, and then sends its model update back
+to the master/parameter server, which aggregates all worker's updates.
+
+One problem with this naive architecture is that there might be stragglers --- some workers might take longer to compute their
+updates, and it's inefficient force the remaining workers to wait for the straggler to get its updates in.
+The solution to this straggler problem is to permit some degree of asynchrony --- model parameters are dispatched 
+asynronously to the workers, before all workers' model updates from the previous iterataion are in.
+The trick to making this work is to enforce that the amount of asynchrony is bounded.
+With bounded asynchrony, you can prove convergence theorems.
+For example,  Theorem 5 in [this paper](http://davidwd.org/papers/2015_ESSP_AAAI_Final.pdf) proves that the
```
Comment (alshedivat) [inline _posts/2019-04-29-lecture-28.md]: Please change the reference to `<d-cite></d-cite>` format.

File: _posts/2019-04-29-lecture-28.md [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+There should be standards for interfaces between nuts and bolts so that any algorithm can be used after any data wrangling method.
+
+### Process
+There should be a workbench where people can easily create, experiment, explain and manage their tasks. A graphical interface may be useful for this process.
+
+### Soundness 
+If you have correctly modularized and implemented components, then you guarantee soundness of the system. You can use the theory and guarantees for each component to build the guarantees for the whole system.
+
+## Real-World Example: Texar
+Texar is a toolkit for text generation built on these principles.
```
Comment (alshedivat) [inline _posts/2019-04-29-lecture-28.md]: Please cite Texar using `<d-cite></d-cite>`. Here's a bibtex entry for you:
```
@article{hu2018texar,
  title={Texar: A Modularized, Versatile, and Extensible Toolkit for Text Generation},
  author={Hu, Zhiting and Shi, Haoran and Yang, Zichao and Tan, Bowen and Zhao, Tiancheng and He, Junxian and Wang, Wentao and Qin, Lianhui and Wang, Di and others},
  journal={arXiv preprint arXiv:1809.00794},
  year={2018}
}
```

File: _posts/2019-04-29-lecture-28.md [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+
+
+
+The key idea is that the model should be modular and composable; instead of building from scratch, we should be assembling composable parts.
+
+
+
+## Distributed Machine Learning
+
+You might have noticed that this year was the first SysML conference (last year's SysML was just a workshop).
```
Comment (alshedivat) [inline _posts/2019-04-29-lecture-28.md]: Could you link to http://www.sysml.cc/?

File: _posts/2019-04-29-lecture-28.md [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+Examples of MPMs include distance metric learning, multi class logistic regression, topic models, sparse coding, neural networks, etc.
+In multiclass logistic regression, the rows of the parameter matrix are the coefficient vectors for the different classes; in sparse coding, the rows of the parameter matrix are the basis vectors used to reconstruct the observed data.  
+
+The trick behind the sufficient factor broadcasting framework is that the SGD update in a matrix-parameterized model is low-rank:
+
+$$ \Delta W = u v^T \quad u = \frac{\partial f(W a_i)}{\partial (W a_i)} \quad v = a_i $$
+
+Therefore, rather than passing around $$\Delta W$$, you can just pass around $$u$$ and $$v$$.
+
+Interestingly, it's not only the SGD update that's low-rank: the [SDCA](http://www.jmlr.org/papers/volume14/shalev-shwartz13a/shalev-shwartz13a.pdf) update is also low-rank.
```
Comment (alshedivat) [inline _posts/2019-04-29-lecture-28.md]: Please change the link to `<d-cite></d-cite>` reference.

File: _posts/2019-04-29-lecture-28.md [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+
+[This paper](http://www.cs.cmu.edu/~pengtaox/papers/uai16_sfb.pdf) provides a detailed analysis of the correctness of this approach.
```
Comment (alshedivat) [inline _posts/2019-04-29-lecture-28.md:440]: Please change the link to `<d-cite></d-cite>` reference.

File: _posts/2019-04-29-lecture-28.md [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+
+### Inference explainability
+
+For explainability, it’s not sufficient to just look at the static weights of the model, as different input points may take different paths through the model.
+Therefore, you need to understand what happens as data travels through the model:
+* Which weights are activated by the data
+* Which decision boundaries are close to the data?
+* What bases are activated by the data?
+
+Recent work on [contextual explanation networks](https://arxiv.org/abs/1705.10301) allow you to incorporate explanations directly into the model.  Intuitively, you train a “personalized” model for every data point.
```
Comment (alshedivat) [inline _posts/2019-04-29-lecture-28.md]: Please change the link to `<d-cite></d-cite>` reference.

**Review by alshedivat — CHANGES_REQUESTED** [review]
Thanks for a great set of notes! Please address the minor requested changes, and we'll be ready to merge. Please do so as soon as possible (end of today is the deadline).

**Comment by jmcohen** [comment]
fixed!

**Review by alshedivat — APPROVED** [review]

**Pull request merged**